src/cfml/system/endpoints/S3.cfc

```diff
@@ -0,0 +1,193 @@
+            cfhttpparam(type='header', name='Host', value='#bucket#.s3.amazonaws.com');
+        }
+        return req.responseheader['x-amz-bucket-region'];
+    }
+
+    private function resolveAwsSettings() {
+        var settings = {
+            awsKey: systemSettings.getSystemSetting('AWS_ACCESS_KEY_ID', ''),
+            awsSecretKey: systemSettings.getSystemSetting('AWS_SECRET_ACCESS_KEY', ''),
+            sessionToken: systemSettings.getSystemSetting('AWS_SESSION_TOKEN', ''),
+            defaultRegion: systemSettings.getSystemSetting('AWS_DEFAULT_REGION', ''),
+            profile: systemSettings.getSystemSetting('AWS_PROFILE', 'default'),
+            configFile: systemSettings.getSystemSetting('AWS_CONFIG_FILE', '~/.aws/config'),
+            credentialsFile: systemSettings.getSystemSetting('AWS_SHARED_CREDENTIALS_FILE', '~/.aws/credentials')
+        };
+
+        if (!settings.awsKey.len() || !settings.awsSecretKey.len()) {
+            var credentials = resolveCredentials(settings.credentialsFile, settings.profile);
+            settings.append(credentials);
+        }
+
+        if (!settings.defaultRegion.len()) {
+            var config = loadSettingsFile(settings.configFile);
+            settings.defaultRegion = config[settings.profile].region ?: 'us-east-1';
+        }
+
+        return settings;
+    }
+
+    private function resolveCredentials(credentialsFile, profile) {
+        // check for an aws credentials file for current user
+        var credentials = loadSettingsFile(credentialsFile);
+        try {
+            return {
+                awsKey: credentials[profile].aws_access_key_id,
+                awsSecretKey: credentials[profile].aws_secret_access_key,
+                sessionToken: credentials[profile].aws_session_token ?: ''
+            };
+        } catch(any e) {
+            // pass
+        }
+
+        // check for IAM role
+        try {
+            var req = '';
+            cfhttp(url = getIamRolePath(), timeout = 1, result = 'req');
+            var roleName = req.filecontent;
+            cfhttp(url = getIamRolePath() & roleName, timeout = 1, result = 'req');
+            var data = deserializeJSON( req.filecontent );
+            return {
+                awsKey: data.AccessKeyId,
+                awsSecretKey: data.SecretAccessKey,
+                sessionToken: data.Token,
+                expires: parseDateTime(data.Expiration)
+            }
+        } catch(any e) {
+            // pass
+        }
+
+        // Credentials unable to be located
+        throw(
+            'Could not locate S3 Credentials',
+            'endpointException'
+        );
+    }
+
+    private function loadSettingsFile(settingsFilePath) {
+        var settings = {};
+        var fullPath = fileSystemUtil.resolvePath(settingsFilePath);
+        if (fileExists(fullPath)) {
+            var settingLines = fileRead(fullPath).listToArray(chr(10));
```

Comment:
This appears to be an INI file format.  Won't this do the trick?
https://cfdocs.org/getprofilestring

Reply:
Nice! I did not know about this function 😃

Reply:
It's ironic that CFML doesn't have any native support for things like YAML, but we do support .ini files.  Go figure :)